src/Razor/test/Microsoft.AspNetCore.Razor.Test.Common/Language/SyntaxTreeVerifier.cs

```diff
@@ -10,12 +10,32 @@ namespace Microsoft.AspNetCore.Razor.Language
     // Verifies recursively that a syntax tree has no gaps in terms of position/location.
     internal class SyntaxTreeVerifier
     {
-        public static void Verify(RazorSyntaxTree syntaxTree)
+        public static void Verify(RazorSyntaxTree syntaxTree, bool ensureFullFidelity = true)
         {
-            new Verifier(syntaxTree.Source).Visit(syntaxTree.Root);
+            var verifier = new Verifier(syntaxTree.Source);
+            verifier.Visit(syntaxTree.Root);
+
+            if (ensureFullFidelity)
+            {
+                var syntaxTreeString = syntaxTree.Root.ToFullString();
+                if (syntaxTree.Source.Length != syntaxTreeString.Length)
```

Comment:
For this and the section below I'd convert the source document to a string and use `Assert.Equal`. Assert.Equal it's smart about giving snippets of a string that don't match which will be easier to diagnose than a single character mismatch.

Reply:
+1

Reply:
Yeah, being about to `recover` the source from the syntax tree is a stronger validation, and would be super duper to have in our tests (if we can).